README.md
LGS Watch (Watchface for Pebble Watch)
=========

Features:

* Display time in either 12h or 24h format
* Day of the week indication
* Configurable through the Pebble app on your smartphone
* Display battery and bluetooth status information (configurable, default On)
* Supporting 11 different languages (configurable, default English)
	- Croatian
	- Dutch
	- English
	- Finnish
	- French
	- German
	- Greek
	- Italian
	- Norwegian
	- Spanish
	- Turkish
* Second date information line (configurable, default Week)

Compiled version of this watchface can be downloaded [here](http://sharedmemorydump.net/pebble-lgs-watchface)  
The font used is Chicago, it is freely available for download [here](http://ttfonts.net/font/5767_Chicago.htm)

## LGS Watch

| LGS Watch               | Information                                     |
|:-----------------------:|:------------------------------------------------|
| ![LGS Watch][lgs_watch] | 1 - 12h or 24 format, based on you Pebble setup<br/>2 - Day of the week indicator<br/>3 - Battery and Bluetooth status information<br/>4 - Date<br/>5 - Second date line |

## Bluetooth Connection Status

| Display                                 | Information                  |
|:---------------------------------------:|:-----------------------------|
| ![](/resources/images/bluetooth.png)    | Connected to smartphone      |
| ![](/resources/images/no_bluetooth.png) | Disconnected from smartphone |

## Battery Status

| Display                                  | Information |
|:----------------------------------------:|:------------|
| ![](/resources/images/batt_charging.png) | Charging    |
| ![](/resources/images/batt_080_100.png)  | 80-100%     |
| ![](/resources/images/batt_060_080.png)  | 60-80%      |
| ![](/resources/images/batt_040_060.png)  | 40-60%      |
| ![](/resources/images/batt_020_040.png)  | 20-40%      |
| ![](/resources/images/batt_000_020.png)  | 0-20%       |

## Configure LGS Watch using your smartphone

| Pebble app on smartphone                | LGS Watch settings                 |
|:---------------------------------------:|:----------------------------------:|
| ![Pebble app on smartphone][pebble_app] | ![LGS Watch settings ][lgs_config] |

## Build Instructions

Clone repository:

	git clone https://github.com/fmeus/pebble-lgswatch.git

Configure and build:

	src/makeall.sh

Compiled version can be found in the folder `releases`

[pebble_app]: /screenshots/pebble_app.png
[lgs_config]: /screenshots/lgs_config.png
[lgs_watch]: /screenshots/lgs_watch.png

changelog.md
v2.4 (Jan 7)

* Shortened monthnames for Finnish

v2.3 (Dec 15)

* Optimized code
* Adjust font size for date/week information to match content
* Added support for additional langauges: Turkish, Greek

v2.2 (Dec 2)

* LGS Watch is now configurable (one version to rule them all) through the Pebble app on your smartphone
* Added Finnish language
* Optimized code

v2.1 (Nov 21)

* Added battery and bluetooth status information

v2.0 (Nov 18)

* Updated for Pebble SDK 2

v1.5 (May 18)

* Added support for Norwegian
* Cleanup of internal code

v1.4 (May 5)

* Added support for Italian
* Added version displaying dd/mm/yy on the last line
* Added version displaying mm/dd/yy on the last line
* Added version displaying Wxx Dxxx (Week and Day of the Year) on the last line

v1.3 (May 2)

* In 12h-mode leading zero for hours has been removed
* Added support for multiple languages (English, Dutch, German, French, Croatian, Spanish)
* Corrected alignment of AM/PM
* Optimized code

v1.2 (Apr 29)

* Adjusted background image so that border is always showing (regardless of viewing angle)
* If 12h-mode leading zero for hours is removed
* Optimized code

v1.1 (Apr 29)

* Updated menu icon, it now shows the actual watch face
* Fixed possible memory issue (unloading custom fonts)
* Hour/minute Separator no longer blinks (to preserve power)

v1.0 (Apr 28)

* Initial release

src/lgs_languages.h
/*
  Project     : LGS Watch watchface
  Copyright   : Copyright (c) 2011-2013 Little Gem Software. All rights reserved.
  Pebble SDK  : 2.0-BETA2
*/

#pragma once

// Some handy macros
#define ConstantGRect(x, y, w, h) {{(x), (y)}, {(w), (h)}}
#define MAX(x, y) (((x) > (y)) ? (x) : (y))


// Forward declarations
void update_time();


// Day line information
static const char *day_lines[] = { "Mo Tu We Th Fr Sa Su"
                                 , "Ma Di Wo Do Vr Za Zo"
                                 , "Mo Di Mi Do Fr Sa So"
                                 , "Lu Ma Me Je Ve Sa Di"
                                 , "Po Vt Sr Ce Pe Su Ne"
                                 , "Lu Ma Mi Ju Vi Sá Do"
                                 , "Lu Ma Me Gi Ve Sa Do"
                                 , "Ma Ti On To Fr Lø Sø"
                                 , "Ma Ti Ke To Pe La Su"
                                 , "Pa Sa Ça Pe Cu Cu Pa"
                                 , "De Tr Te Pé Pa Sá Ku"
                                 };


// GRect for day highlights (fixed y=78, h=11)
typedef struct GBox { int16_t x; int16_t w; } GBox;
#define GBox(x, w) {(x), (w)}
static const GBox highlight_box[][7] = { { GBox( 4, 20 ), GBox( 26, 16 ), GBox( 46, 20 ), GBox( 68, 16 ), GBox( 88, 14 ), GBox( 105, 15 ), GBox( 124, 15 ) }
                                       , { GBox( 3, 20 ), GBox( 27, 12 ), GBox( 43, 20 ), GBox( 67, 16 ), GBox( 87, 15 ), GBox( 105, 16 ), GBox( 125, 16 ) }
                                       , { GBox( 6, 20 ), GBox( 30, 12 ), GBox( 46, 16 ), GBox( 66, 16 ), GBox( 86, 14 ), GBox( 103, 15 ), GBox( 122, 15 ) }
                                       , { GBox( 3, 15 ), GBox( 22, 20 ), GBox( 46, 20 ), GBox( 69, 16 ), GBox( 89, 16 ), GBox( 109, 15 ), GBox( 128, 12 ) }
                                       , { GBox( 6, 16 ), GBox( 26, 14 ), GBox( 44, 14 ), GBox( 61, 16 ), GBox( 81, 16 ), GBox( 101, 15 ), GBox( 120, 17 ) }
                                       , { GBox( 5, 15 ), GBox( 24, 20 ), GBox( 48, 16 ), GBox( 67, 16 ), GBox( 87, 12 ), GBox( 103, 15 ), GBox( 122, 16 ) }
                                       , { GBox( 3, 15 ), GBox( 22, 20 ), GBox( 46, 20 ), GBox( 70, 12 ), GBox( 86, 16 ), GBox( 106, 15 ), GBox( 125, 16 ) }
                                       , { GBox( 8, 20 ), GBox( 30, 12 ), GBox( 46, 16 ), GBox( 64, 16 ), GBox( 84, 14 ), GBox( 101, 15 ), GBox( 120, 15 ) }
                                       , { GBox( 6, 20 ), GBox( 28, 12 ), GBox( 44, 17 ), GBox( 63, 16 ), GBox( 83, 16 ), GBox( 103, 15 ), GBox( 122, 15 ) }
                                       , { GBox( 4, 16 ), GBox( 24, 15 ), GBox( 43, 16 ), GBox( 63, 16 ), GBox( 83, 16 ), GBox( 103, 16 ), GBox( 123, 16 ) }
                                       , { GBox( 7, 16 ), GBox( 25, 15 ), GBox( 41, 16 ), GBox( 61, 16 ), GBox( 81, 16 ), GBox( 101, 15 ), GBox( 120, 17 ) }
                                       };


// Month names
static const char *month_names_arr[][12] = { { "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec" }
                                           , { "jan", "feb", "mrt", "apr", "mei", "jun", "jul", "aug", "sep", "okt", "nov", "dec" }
                                           , { "Jan", "Feb", "März", "Apr", "Mai", "Jun", "Jul", "Aug", "Sep", "Okt", "Nov", "Dez" }
                                           , { "janv", "févr", "mars", "avr", "mai", "juin", "juil", "août", "sept", "oct", "nov", "déc" }
                                           , { "siječanj", "veljača", "ožujak", "travanj", "svibanj", "lipanj", "srpanj", "kolovoz", "rujan", "listopad", "studeni", "prosinac" }
                                           , { "enero", "feb", "marzo", "abr", "mayo", "jun", "jul", "agosto", "sept", "oct", "nov", "dic" }
                                           , { "genn", "febbr", "mar", "apr", "magg", "giugno", "luglio", "ag", "sett", "ott", "nov", "dic" }
                                           , { "Jan", "Feb", "Mar", "Apr", "Mai", "Jun", "Jul", "Aug", "Sep", "Okt", "Nov", "Des" }
                                           , { "Tammi", "Helmi", "Maalis", "Huhti", "Touko", "Kesä", "Heinä", "Elo", "Syys", "Loka", "Marras", "Joulu" }
                                           , { "Ocak", "Şubat", "Mart", "Nisan", "Mayıs", "Haziran", "Temmuz", "Ağustos", "Eylül", "Ekim", "Kasım", "Aralık" }
                                           , { "Ianuários", "Fevruários", "Mártios", "Aprílios", "Máios", "Iúnios", "Iúlios", "Avghustos", "Septémvrios", "Októvrios", "Noémvrios", "Thekémvrios" }
                                           };


// Format week
static const char *week_formats[] = { "Week %V"
                                    , "Week %V"
                                    , "Woche %V"
                                    , "Semaine %V"
                                    , "%V. tjedan"
                                    , "%V. semana"
                                    , "%V. Setti."
                                    , "Uke %V"
                                    , "VKO %V"
                                    , "Hafta %V"
                                    , "Evdomáda %V"
                                    };


// Format date
static const char *date_formats[] = { "%b %e"
                                    , "%e "
                                    , "%e. "
                                    , "%e "
                                    , "%e. "
                                    , "%e. "
                                    , "%e. "
                                    , "%e. "
                                    , "%e. "
                                    , "%e "
                                    , "%e "
                                    };


// Formats for second line
static const char *alt_formats[] = { "<See date_formats[]>"
                                   , "%Y-%j"
                                   , "%d/%m/%y"
                                   , "%m/%d/%y"
                                   , "W%V D%j"
                                   };


static char week_text[] = "               ";
static char date_text[] = "               ";
static char alt_text[]  = "               ";


// Ordinal numbers
static const char *ordinal_numbers[] = { "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th"
                                       , "th", "th", "th", "th", "th", "th", "th", "th", "th", "th"
                                       , "st", "nd", "rd", "th", "th", "th", "th", "th", "th", "th"
                                       , "st" };

src/js/pebble-js-app.js
var initialized = false;


Pebble.addEventListener( "ready", function(e) {
	initialized = true;
});


Pebble.addEventListener( "appmessage", function(e) {
	localStorage.setItem( "status", e.payload.status );
	localStorage.setItem( "language", e.payload.language );
	localStorage.setItem( "format", e.payload.format );
});


Pebble.addEventListener( "showConfiguration", function(e) {
	Pebble.openURL( "http://fmeus.github.io/www/pebble_lgs/lgs_watch-2.3.html" + 
		            "?status=" + localStorage.getItem( "status" ) + 
		            "&language=" + localStorage.getItem( "language" ) + 
		            "&format=" + localStorage.getItem( "format" ) );
});


Pebble.addEventListener( "webviewclosed", function(e) {
	if ( ( typeof( e.response ) == 'string') && ( e.response.length > 0 ) ) {
		try {
			var options = JSON.parse( e.response );
			localStorage.setItem( "status", options.status );
			localStorage.setItem( "language", options.language );
			localStorage.setItem( "format", options.format );
			Pebble.sendAppMessage( options );
		} catch(e) {
		}
	}
});

src/lgs.c
/*
  Project     : LGS Watch watchface
  Copyright   : Copyright (c) 2011-2013 Little Gem Software. All rights reserved.
  Pebble SDK  : 2.0-BETA2
*/


#include <pebble.h>
#include "lgs_languages.h"


// Setting values
static enum statusKeys { STATUS_OFF = 0
                       , STATUS_ON
                       } current_status;
static enum formatKeys { FORMAT_WEEK = 0
                       , FORMAT_DOTY
                       , FORMAT_DDMMYY
                       , FORMAT_MMDDYY
                       , FORMAT_WXDX
                       } current_format;
static enum languageKeys { LANG_EN = 0
                         , LANG_NL
                         , LANG_DE
                         , LANG_FR
                         , LANG_HR
                         , LANG_ES
                         , LANG_IT
                         , LANG_NO
                         , LANG_FI
                         , LANG_TR
                         , LANG_GR
                         } current_language;


// Setting keys
enum settingKeys {
  SETTING_STATUS_KEY = 0x0,
  SETTING_LANGUAGE_KEY = 0x1,
  SETTING_FORMAT_KEY = 0x2
};


// All UI elements
static Window         *window;
static GBitmap        *background_image;
static GBitmap        *battery_image;
static GBitmap        *bluetooth_image;
static BitmapLayer    *background_layer;
static BitmapLayer    *battery_layer;
static BitmapLayer    *bluetooth_layer;
static TextLayer      *text_time_layer;
static TextLayer      *text_ampm_layer;
static TextLayer      *text_days_layer;
static TextLayer      *text_date_layer;
static TextLayer      *text_week_layer;
static GFont          *font_time;
static GFont          *font_days;
static GFont          *font_date;
static GFont          *font_date_small;
static InverterLayer  *currentDayLayer;
static AppSync        app;
static uint8_t        sync_buffer[256];


// Define layer rectangles (x, y, width, height)
GRect TIME_RECT      = ConstantGRect(   9,   8, 124,  45 );
GRect AMPM_RECT      = ConstantGRect(  51,  55,  40,  16 );
GRect DATE_RECT      = ConstantGRect(   0, 105, 144,  25 );
GRect WEEK_RECT      = ConstantGRect(   0, 132, 144,  25 );
GRect DAYS_RECT      = ConstantGRect(   0,  76, 144,  13 );
GRect BATT_RECT      = ConstantGRect( 123,  94,  17,   9 );
GRect BT_RECT        = ConstantGRect(   4,  94,  17,   9 );
GRect OFF_DATE_RECT  = ConstantGRect(   0, 100, 144,  25 );
GRect OFF_WEEK_RECT  = ConstantGRect(   0, 130, 144,  25 );


// Define placeholders for time and date
static char time_text[] = "00:00";
static char ampm_text[] = "AM";


// Previous bluetooth connection status
static bool prev_bt_status = false;


/*
  Setup new TextLayer
*/
static TextLayer * setup_text_layer( GRect rect, GTextAlignment align , GFont font ) {
  TextLayer *newLayer = text_layer_create( rect );
  text_layer_set_text_color( newLayer, GColorWhite );
  text_layer_set_background_color( newLayer, GColorClear );
  text_layer_set_text_alignment( newLayer, align );
  text_layer_set_font( newLayer, font );

  return newLayer;
}


/*
  Handle bluetooth events
*/
void handle_bluetooth( bool connected ) {
  if ( bluetooth_image != NULL ) {
    gbitmap_destroy( bluetooth_image );
  }

  if ( connected ) {
    bluetooth_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BLUETOOTH );
  } else {
    bluetooth_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_NO_BLUETOOTH );
    if ( prev_bt_status != connected ) {
      vibes_short_pulse();
      light_enable_interaction();
    }
  }

  prev_bt_status = connected;

  bitmap_layer_set_bitmap( bluetooth_layer, bluetooth_image );
}


/*
  Handle battery events
*/
void handle_battery( BatteryChargeState charge_state ) {
  if ( battery_image != NULL ) {
    gbitmap_destroy( battery_image );
  }

  if ( charge_state.is_charging ) {
    battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_CHARGING );
  } else {
    switch ( charge_state.charge_percent ) {
      case 0 ... 20:
        battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_000_020 );
        break;
      case 21 ... 40:
        battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_020_040 );
        break;
      case 41 ... 60:
        battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_040_060 );
        break;
      case 61 ... 80:
        battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_060_080 );
        break;
      case 81 ... 100:
        battery_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BATT_080_100 );
        break;
      }
  }

  bitmap_layer_set_bitmap( battery_layer, battery_image );
}


/*
  Update status indicators for the battery and bluetooth connection
*/
void update_status( void ) {
  handle_battery( battery_state_service_peek() );
  handle_bluetooth( bluetooth_connection_service_peek() );
}


/*
  Build day line based on user selected first day of the week
*/
static void set_day_line( void ) {
  text_layer_set_text( text_days_layer, day_lines[current_language] );  
}


/*
  Handle update in settings
*/
static void tuple_changed_callback( const uint32_t key, const Tuple* tuple_new, const Tuple* tuple_old, void* context ) {
  int value = tuple_new->value->uint8;  

  // APP_LOG( APP_LOG_LEVEL_DEBUG, "Update for key = %lu", key );

  switch ( key ) {
    case SETTING_STATUS_KEY:
      // APP_LOG( APP_LOG_LEVEL_DEBUG, "       Setting = Status" );

      persist_write_int( SETTING_STATUS_KEY, value );
      current_status = value;

      // APP_LOG( APP_LOG_LEVEL_DEBUG, "        Stored = %lu", persist_read_int( SETTING_STATUS_KEY ) );

      // Reposition status layers
      layer_set_frame( bitmap_layer_get_layer( battery_layer ), ( current_status == STATUS_ON ) ? BATT_RECT : GRectZero );
      layer_set_frame( bitmap_layer_get_layer( bluetooth_layer ), ( current_status == STATUS_ON ) ? BT_RECT : GRectZero );

      // Repasition date and week layers
      layer_set_frame( text_layer_get_layer( text_date_layer ), ( current_status == STATUS_ON ) ? DATE_RECT : OFF_DATE_RECT );
      layer_set_frame( text_layer_get_layer( text_week_layer ), ( current_status == STATUS_ON ) ? WEEK_RECT : OFF_WEEK_RECT );
      break;

    case SETTING_LANGUAGE_KEY:
      // APP_LOG( APP_LOG_LEVEL_DEBUG, "       Setting = Language" );

      persist_write_int( SETTING_LANGUAGE_KEY, value );
      current_language = value;

      // APP_LOG( APP_LOG_LEVEL_DEBUG, "        Stored = %lu", persist_read_int( SETTING_LANGUAGE_KEY ) );

      // Update day line
      set_day_line();

      // Reset font for date and week layers
      text_layer_set_font( text_date_layer, font_date );
      text_layer_set_font( text_week_layer, font_date );
      break;

    case SETTING_FORMAT_KEY:
      // APP_LOG( APP_LOG_LEVEL_DEBUG, "       Setting = Format" );

      persist_write_int( SETTING_FORMAT_KEY, value );
      current_format = value;

      // APP_LOG( APP_LOG_LEVEL_DEBUG, "        Stored = %lu", persist_read_int( SETTING_FORMAT_KEY ) );

      break;
  }

  // Refresh display
  update_time();
}


/*
  Handle errors
*/
static void app_error_callback( DictionaryResult dict_error, AppMessageResult app_message_error, void* context ) {
  APP_LOG( APP_LOG_LEVEL_DEBUG, "app error %d", app_message_error );
  vibes_double_pulse();
}


GRect box_to_rect( GBox box ) {
  return GRect( box.x, 78, box.w, 11 );
}


/*
  Handle tick events
*/
void handle_tick( struct tm *tick_time, TimeUnits units_changed ) {
  // Update text layer for current day
  int today = tick_time->tm_wday - 1; if ( today < 0 ) { today = 6; }
  layer_set_frame( inverter_layer_get_layer( currentDayLayer ), box_to_rect( highlight_box[current_language][today] ) );

  strftime( date_text, sizeof( date_text ), date_formats[current_language], tick_time );
  if ( current_language == LANG_EN ) {
    strcat( date_text, ordinal_numbers[tick_time->tm_mday - 1] );
  } else {
    strcat( date_text, month_names_arr[current_language][tick_time->tm_mon] );
  }
  text_layer_set_text( text_date_layer, date_text );

  // Update week or day of the year (i.e. Week 15 or 2013-118)
  if ( current_format == FORMAT_WEEK ) {
    strftime( week_text, sizeof( week_text ), week_formats[current_language], tick_time );
    text_layer_set_text( text_week_layer, week_text );
  } else {
    strftime( alt_text, sizeof( alt_text ), alt_formats[current_format], tick_time );
    text_layer_set_text( text_week_layer, alt_text );
  }

  // Determine size of week and date layers
  GSize weekSize = text_layer_get_content_size( text_week_layer );
  GSize dateSize = text_layer_get_content_size( text_date_layer );

  // Switch to a smaller font if content doesn't fit
  if ( MAX( weekSize.w, dateSize.w ) > 135 ) {
    text_layer_set_font( text_date_layer, font_date_small );
    text_layer_set_font( text_week_layer, font_date_small );
  }

  // Display hours (i.e. 18 or 06)
  strftime( time_text, sizeof( time_text ), clock_is_24h_style() ? "%H:%M" : "%I:%M", tick_time );

  // Remove leading zero (only in 12h-mode)
  if ( !clock_is_24h_style() && (time_text[0] == '0') ) {
    memmove( time_text, &time_text[1], sizeof( time_text ) - 1 );
  }
  text_layer_set_text( text_time_layer, time_text );

  // Update AM/PM indicator (i.e. AM or PM or nothing when using 24-hour style)
  strftime( ampm_text, sizeof( ampm_text ), clock_is_24h_style() ? "" : "%p", tick_time );
  text_layer_set_text( text_ampm_layer, ampm_text );

  // Update status information 
  if ( current_status == STATUS_ON ) {
    update_status();
  }
}


/*
  Force update of time
*/
void update_time() {
  // Get current time
  time_t now = time( NULL );
  struct tm *tick_time = localtime( &now );

  // Force update to avoid a blank screen at startup of the watchface
  handle_tick( tick_time, SECOND_UNIT );
}


/*
  Initialization
*/
void handle_init( void ) {
  window = window_create();
  window_stack_push( window, true );
  Layer *window_layer = window_get_root_layer( window );

  // Read persistent data
  if ( persist_exists( SETTING_STATUS_KEY ) ) {
    current_status = persist_read_int( SETTING_STATUS_KEY );
  } else {
    current_status = STATUS_ON;
  }
  if ( persist_exists( SETTING_LANGUAGE_KEY ) ) {
    current_language = persist_read_int( SETTING_LANGUAGE_KEY );
  } else {
    current_language = LANG_EN;
  }
  if ( persist_exists( SETTING_FORMAT_KEY ) ) {
    current_format = persist_read_int( SETTING_FORMAT_KEY );
  } else {
    current_format = FORMAT_WEEK;
  }

  // Read watchface settings from persistent data or use default values
  current_status = persist_exists( SETTING_STATUS_KEY ) ? persist_read_int( SETTING_STATUS_KEY ) : STATUS_ON;
  current_language = persist_exists( SETTING_LANGUAGE_KEY ) ? persist_read_int( SETTING_LANGUAGE_KEY ) : LANG_EN;
  current_format = persist_exists( SETTING_FORMAT_KEY ) ? persist_read_int( SETTING_FORMAT_KEY ) : FORMAT_WEEK;

  // APP_LOG( APP_LOG_LEVEL_DEBUG, "persist_read_int for Status = %lu", persist_read_int( SETTING_STATUS_KEY ) );
  // APP_LOG( APP_LOG_LEVEL_DEBUG, "persist_read_int for Langauge = %lu", persist_read_int( SETTING_LANGUAGE_KEY ) );
  // APP_LOG( APP_LOG_LEVEL_DEBUG, "persist_read_int for Format = %lu", persist_read_int( SETTING_FORMAT_KEY ) );

  // Background image
  background_image = gbitmap_create_with_resource( RESOURCE_ID_IMAGE_BACKGROUND );
  background_layer = bitmap_layer_create( layer_get_frame( window_layer ) );
  bitmap_layer_set_bitmap( background_layer, background_image );
  layer_add_child( window_layer, bitmap_layer_get_layer( background_layer ) );

  // Initial settings
  Tuplet initial_values[] = { TupletInteger( SETTING_STATUS_KEY, current_status )
                            , TupletInteger( SETTING_LANGUAGE_KEY, current_language )
                            , TupletInteger( SETTING_FORMAT_KEY, current_format )
                            };

  // Open AppMessage to transfers
  app_message_open( 160, 160 );

  // Initialize AppSync
  app_sync_init( &app, sync_buffer
               , sizeof( sync_buffer )
               , initial_values
               , ARRAY_LENGTH( initial_values )
               , tuple_changed_callback
               , app_error_callback
               , NULL 
               );

  // Perform sync
  app_sync_set( &app, initial_values, ARRAY_LENGTH( initial_values ) );

  // Adjust GRect for Hours, Minutes and Blink to compensate for missing AM/PM indicator
  if ( clock_is_24h_style() ) {
    TIME_RECT.origin.y = TIME_RECT.origin.y + 6;
  }

  // Load fonts
  font_time = fonts_load_custom_font( resource_get_handle( RESOURCE_ID_FONT_CHICAGO_SUBSET_40 ) );
  font_days = fonts_load_custom_font( resource_get_handle( RESOURCE_ID_FONT_CHICAGO_12 ) );
  font_date_small = fonts_load_custom_font( resource_get_handle( RESOURCE_ID_FONT_CHICAGO_14 ) );
  font_date = fonts_load_custom_font( resource_get_handle( RESOURCE_ID_FONT_CHICAGO_20 ) );

  // Setup time layer
  text_time_layer = setup_text_layer( TIME_RECT, GTextAlignmentCenter, font_time );
  layer_add_child( window_layer, text_layer_get_layer( text_time_layer ) );

  // Setup AM/PM name layer
  text_ampm_layer = setup_text_layer( AMPM_RECT, GTextAlignmentCenter, font_days );
  layer_add_child( window_layer, text_layer_get_layer( text_ampm_layer ) );

  // Setup days line layer
  text_days_layer = setup_text_layer( DAYS_RECT, GTextAlignmentCenter, font_days );
  layer_add_child( window_layer, text_layer_get_layer( text_days_layer ) );

  // Setup date layer
  text_date_layer = setup_text_layer( ( current_status == STATUS_ON ) ? DATE_RECT : OFF_DATE_RECT
                                    , GTextAlignmentCenter
                                    , font_date );
  layer_add_child( window_layer, text_layer_get_layer( text_date_layer ) );

  // Setup week layer
  text_week_layer = setup_text_layer( ( current_status == STATUS_ON ) ? WEEK_RECT : OFF_WEEK_RECT
                                    , GTextAlignmentCenter
                                    , font_date );
  layer_add_child( window_layer, text_layer_get_layer( text_week_layer ) );

  // Setup battery layer
  battery_layer = bitmap_layer_create( ( current_status == STATUS_ON ) ? BATT_RECT : GRectZero );
  layer_add_child( window_layer, bitmap_layer_get_layer( battery_layer ) );

  // Setup bluetooth layer
  bluetooth_layer = bitmap_layer_create( ( current_status == STATUS_ON ) ? BT_RECT : GRectZero );
  layer_add_child( window_layer, bitmap_layer_get_layer( bluetooth_layer ) );

  // Add inverter layer (indicator for the current day of the week)
  currentDayLayer = inverter_layer_create( GRectZero );
  layer_add_child( window_layer, inverter_layer_get_layer( currentDayLayer ) );

  // Subscribe to services
  tick_timer_service_subscribe( SECOND_UNIT, handle_tick );

  // Force update to avoid a blank screen at startup of the watchface
  update_time();
}


/*
  Destroy GBitmap and BitmapLayer
*/
void destroy_graphics( GBitmap *image, BitmapLayer *layer ) {
  layer_remove_from_parent( bitmap_layer_get_layer( layer ) );
  bitmap_layer_destroy( layer );
  if ( image != NULL ) {
    gbitmap_destroy( image );
  }
}


/*
  dealloc
*/
void handle_deinit( void ) {
  // Unsubscribe from services
  tick_timer_service_unsubscribe();

  // Destroy image objects
  destroy_graphics( background_image, background_layer );
  destroy_graphics( battery_image, battery_layer );
  destroy_graphics( bluetooth_image, bluetooth_layer );

  // Destroy text objects
  text_layer_destroy( text_time_layer );
  text_layer_destroy( text_ampm_layer );
  text_layer_destroy( text_days_layer );
  text_layer_destroy( text_date_layer );
  text_layer_destroy( text_week_layer );

  // Destroy inverter layer
  inverter_layer_destroy( currentDayLayer );
  
  // Destroy font objects
  fonts_unload_custom_font( font_time );
  fonts_unload_custom_font( font_days );
  fonts_unload_custom_font( font_date_small );
  fonts_unload_custom_font( font_date );

  // Clean up AppSync system
  app_sync_deinit( &app );

  // Destroy window
  window_destroy( window );
}


/*
  Main process
*/
int main( void ) {
  handle_init();
  app_event_loop();
  handle_deinit();
}